extensions/ql-vscode/src/telemetry.ts

```diff
@@ -0,0 +1,159 @@
+import { Disposable, ConfigurationTarget, Extension, ExtensionContext, workspace } from 'vscode';
+import TelemetryReporter from 'vscode-extension-telemetry';
+import { CANARY_FEATURES, ENABLE_TELEMETRY, LOG_TELEMETRY } from './config';
+import * as appInsights from 'applicationinsights';
+import { logger } from './logging';
+import { UserCancellationException } from './commandRunner';
+import { showBinaryChoiceDialog } from './helpers';
+
+// Key is injected at build time through the APP_INSIGHTS_KEY environment variable.
+const key = 'REPLACE-APP-INSIGHTS-KEY';
+
+let reporter: TelemetryReporter | undefined;
+let listener: Disposable | undefined;
+
+export enum CommandCompletion {
+  Success = 'Success',
+  Failed = 'Failed',
+  Cancelled = 'Cancelled'
+}
+
+export async function initializeTelemetry(extension: Extension<any>, ctx: ExtensionContext): Promise<void> {
+  await requestTelemetryPermission(ctx);
+
+  registerListener(extension, ctx);
+  if (reporter) {
+    reporter.dispose();
+    reporter = undefined;
+  }
+  if (ENABLE_TELEMETRY.getValue<boolean>()) {
```

Comment:
Should we be checking the global telemetry setting too, as stated in the writeup?
For consistency, I suggest adding an `isTelemetryEnabled` helper to `config.ts`, which checks the relevant settings, and always using that. (We could even enforce it by query.)

Reply:
No. The global telemetry value is checked implicitly in the `vscode-extension-telemetry` package.

Reply:
I started working on a PR so that this package can do all the checking of both settings, but the maintainer is not responsive. I created a very simple PR as a trial balloon, and there was no response.

Reply:
> Should we be checking the global telemetry setting too, as stated in the writeup?
> For consistency, I suggest adding an `isTelemetryEnabled` helper to `config.ts`, which checks the relevant settings, and always using that. (We could even enforce it by query.)

I don't think this is necessary since the check is a one-liner anyway.